### DISCUSSION TITLE
How to prebundle external python packages?

### DISCUSSION THREAD

**Discussion by rheanV** [pyodide/pyodide#4575] [Q&A]
I have a use case where my app will be in the internal network without access to the internet. I want to bundle a couple of python packages like tinyHtml. How do I achieve that? Where should I add the wheels? I use pyodide NPM package.

**Comment by ryanking13** [comment]
If your packages are pure python packages, you can add your package into `python_stdlib.zip` (Find it under `node_modules/pyodide`)

1. copy python_stdlib.zip and extract it somewhere.
2. add your packages to unpacked python_stdlib directory
3. repack the directory into a zip file (e.g. `python_stdlib_patched.zip`)
4. when loading Pyodide, set `stdLibURL` parameter to point the zipfile you just created.
5. you can use the packages in the zipfile as if they are stdlib modules.

```sh
unzip python_stdlib.zip -d python_stdlib

mkdir python_stdlib/mymodule
echo "def myfunc():" > python_stdlib/mymodule/run.py
echo "    print('Hello from mymodule.myfunc()')" >> python_stdlib/mymodule/run.py

cd python_stdlib && zip -r python_stdlib_patched.zip . && mv python_stdlib_patched.zip ../ && cd ..
rm -rf python_stdlib
```

```mjs
import { loadPyodide } from 'pyodide'

async function main() {
    const pyodide = await loadPyodide({ stdLibURL: "./python_stdlib_patched.zip" })
    pyodide.runPython(`
import mymodule.run
mymodule.run.myfunc()
`)
}

main()

**Reply by kkoomen** [reply]
@ryanking13 how can this be done with modules like pytest and bundle this into the stdlib? because I don't want to be dependent of external URLs. Can you provide an example how to bundle pytest into stdlib?

**Reply by ryanking13** [reply]
Download pytest wheel from PyPI, uncompress it (wheel is just a zipfile) and put it under the zipfile. 

```bash
wget https://files.pythonhosted.org/packages/4d/7e/c79cecfdb6aa85c6c2e3cf63afc56d0f165f24f5c66c03c695c4d9b84756/pytest-8.1.1-py3-none-any.whl
unzip pytest-8.1.1-py3.none-any.whl
[...put uncompressed files into the zipfile]
```

**Comment by rheanV** [comment]
@ryanking13 I get this error when I bundle the libraries together and create a new python_stdlibs.zip. 
![IMG_8857](https://github.com/pyodide/pyodide/assets/42920310/08697bae-3d2d-4f3d-8ba0-d70665682e07)

 This is what my python_stdlibs looks like. 
![IMG_8858](https://github.com/pyodide/pyodide/assets/42920310/1884f0bb-866c-47c9-bb8c-a100fdf24141)

I make sure to zip this and set the stdLibURL parameter to point to the zip file.

**Reply by ryanking13** [reply]
Could you please share the file you used, so I can check?

**Reply by rheanV** [reply]
So which file are you referring to? Pyodide.asm.js or my code file? Basically, I downloaded the external Pyodide library(I am using version 0.25.1)  and replaced the python_stdlib with the new zip file. Here's how I call and load the library.
![IMG_8864](https://github.com/pyodide/pyodide/assets/42920310/47781c6c-2ea5-42a0-b8cd-c82b0d49c231)

**Reply by ryanking13** [reply]
> So which file are you referring to? 

stdlib.zip file you used.

**Comment by siefkenj** [comment]
I was able to do this in another way. I used `vite` to bundle the raw `.whl` files as `Uint8Array`s. Then I unpacked them with `unpackArchive`. After that, they could be imported by `runPythonAsync`! If you use a `.tar.gz` instead of a `.whl`, you need to add the sub-directory to the path (as you can see with my `pretext_wasm` package.

An example:
```typescript
import { loadPyodide } from "pyodide";

// @ts-ignore
import rawZipImport from "../python/dist/pretext_wasm-0.1.0.tar.gz?uint8array&base64";
const rawZip = rawZipImport as Uint8Array;
// @ts-ignore
import rawMicropipImport from "./assets/micropip-0.7.1-py3-none-any.whl?uint8array&base64";
const rawMicropip = rawMicropipImport as Uint8Array;
// @ts-ignore
import rawPackagingImport from "./assets/packaging-24.2-py3-none-any.whl?uint8array&base64";
const rawPackaging = rawPackagingImport as Uint8Array;
// @ts-ignore
import rawLxmlImport from "./assets/lxml-5.2.1-cp312-cp312-pyodide_2024_0_wasm32.whl?uint8array&base64";
const rawLxml = rawLxmlImport as Uint8Array;

export async function runPythonCode() {
    console.log("Running Python code");

    const pyodide = await loadPyodide();
    await pyodide.unpackArchive(rawZip, "gztar");
    await pyodide.unpackArchive(rawMicropip, "zip");
    await pyodide.unpackArchive(rawPackaging, "zip");
    await pyodide.unpackArchive(rawLxml, "zip");
    // For some reason this is needed to get pyodide to recognize the package
    await pyodide.runPythonAsync(`
        import sys
        sys.path.append("pretext_wasm-0.1.0")
        `);

    const ret = await pyodide.runPythonAsync(`
        import pretext_wasm
        print(dir(pretext_wasm))
        `);

    console.log("Code ran and returned", ret);
}
```

**Reply by kahrpatrick** [reply]
Thank you! This is actually pretty useful.
I am bundling the wheels as extra resources in an Electron app and load them as buffers using `fs.readFile`.
After trying to build a custom `python_stdlib.zip` with the desired packages I came to the conclusion that this is the best way to do it as I don't have to mess with the pyodide package as such and only need to make sure my CI pipeline builds the packages with the correct pyodide version. 
I cannot understand why there is no official way to vendor packages into the stdlib bundle as dynamically loading them over the Internet doesn't work for applications that run in restricted networks or have strict privacy policy requirements.

**Reply by ryanking13** [reply]
> I cannot understand why there is no official way to vendor packages into the stdlib bundle as dynamically loading them over the Internet doesn't work for applications that run in restricted networks or have strict privacy policy requirements.

I think the main reason we can't say just " put the packages into the zip file" is that putting non-pure Python packages into zipfile will not work, at least for now. Also, packages that depend on the relative path of other files will not work. So there are some common and tricky cases that cannot be covered with the zipfile.

**Reply by kahrpatrick** [reply]
Sure. I do understand that it is difficult to make this work for all packages. I am talking about packages like `numpy` that are already part of the distribution but not included in the `python_stdlib.zip`. I am now manually 'sideloading' them as my app needs to work without Internet access.

**Reply by jsalsman** [reply]
I was able to adapt @siefkenj's method to reduce load times for `pandas` and `statsmodels` in a slow client browser from about 80 to 60 seconds, a 25% speedup over using the CDN. I am sure a memory snapshot will be much faster when https://github.com/pyodide/pyodide/issues/5195 gets fixed, but I thought it likely helpful to share the details here, starting with the [build instructions](https://pyodide.org/en/stable/development/building-from-sources.html#building-a-full-pyodide-distribution):

``` bash
git clone --recursive https://github.com/pyodide/pyodide
cp make_preload.py pyodide # see below
cd pyodide
./run_docker
make
git clone https://github.com/pyodide/pyodide-recipes
pyodide build-recipes "pandas, statsmodels" --recipe-dir pyodide-recipes/packages --install
# That takes a little over an hour on a 2-core GitHub Codespace; so use 4 cores
python make_preload.py
exit
rm -rf ../custom-pyodide
cp -r custom-pyodide ..
cd ..
git add custom-pyodide
git commit -m "Update custom pyodide build"
git push
```

Here is `make_preload.py`:

``` python
from pathlib import Path
import shutil, zipfile, tarfile

dist = Path("dist")
out = Path("custom-pyodide")
stage = Path("_stage_pkgs")

# Clean previous builds
for path in [out, stage]:
    if path.exists():
        shutil.rmtree(path)
out.mkdir(parents=True)

# The build produced cp313 wheels; stage under the corresponding path -- NOTE WELL: Check this after 2025
sitepkgs = stage / "lib" / "python3.13" / "site-packages"
sitepkgs.mkdir(parents=True)

def is_testish(name: str) -> bool:
    return name.lower().startswith(("test", "pytest"))

# Extract wheels into site-packages
for whl in dist.glob("*.whl"):
    if not is_testish(whl.name):
        with zipfile.ZipFile(whl) as zf:
            zf.extractall(sitepkgs)

# Expand non-stdlib zips into the staged filesystem root
for z in dist.glob("*.zip"):
    if z.name != "python_stdlib.zip" and not is_testish(z.name):
        with zipfile.ZipFile(z) as zf:
            zf.extractall(stage)

# Ensure shared libraries are in standard library search paths.
for so in list(stage.glob("*.so")):
    shutil.move(so, sitepkgs / so.name)

# Create one unified archive containing everything staged above.
with tarfile.open(out / "packages.tgz", "w:gz") as tf:
    for p in sorted(stage.glob("*")):
        tf.add(p, arcname=p.relative_to(stage))

# Copy the runtime essentials
for name in ["pyodide.js", "pyodide.asm.js", "pyodide.asm.wasm",
             "python_stdlib.zip", "pyodide-lock.json", "pyodide.mjs"]:
    if (src := dist / name).exists():
        shutil.copy2(src, out / name)

shutil.rmtree(stage)
```

In my single-page app, I do this:

``` html
    <script>
        // --- Web Worker Code ---
        // Resolve base URL to ensure worker can find files relative to index.html
        const baseUrl = window.location.href.substring(0, window.location.href.lastIndexOf('/') + 1);

        const workerCode = `
            importScripts("${baseUrl}custom-pyodide/pyodide.js");

            function updateStatus(msg) {
                self.postMessage({ type: 'status', message: msg });
            }
            self.updateStatus = updateStatus;

            async function loadPyodideAndPackages() {
                self.updateStatus("Loading Pyodide...");
                self.pyodide = await loadPyodide({ indexURL: "${baseUrl}custom-pyodide/" });

                self.updateStatus("Unpacking packages...");
                const resp = await fetch("${baseUrl}custom-pyodide/packages.tgz");
                const buf = await resp.arrayBuffer();
                await self.pyodide.unpackArchive(buf, "tgz", { extractDir: "/" });

                self.updateStatus("Importing pandas...");
                await self.pyodide.runPythonAsync("import pandas as pd");
                self.updateStatus("Importing statsmodels...");
                await self.pyodide.runPythonAsync("from statsmodels.tsa.holtwinters import ExponentialSmoothing");
                self.updateStatus("Imported packages...");
            }

            let pyodideReadyPromise = loadPyodideAndPackages();

            const pythonScript = \`
import pandas as pd
from statsmodels.tsa.holtwinters import ExponentialSmoothing
import js
...\`
            self.onmessage = async (event) => {
                await pyodideReadyPromise;
                const { type, payload } = event.data;

                if (type === 'init') {
                    // Initialize Python environment
                    await self.pyodide.runPythonAsync(pythonScript);
...
        // --- Main UI Logic ---
        const blob = new Blob([workerCode], { type: "text/javascript" });
        const worker = new Worker(URL.createObjectURL(blob));
...
        // Start initialization
        worker.postMessage({ type: 'init' });
...
    </script>
```

I hope that helps others looking here to shave off load time and avoid reliance on the internet CDN. The SPA is at https://covid.talknicer.com and in https://github.com/jsalsman/covid-forecast if you want to try it and look at the full implementation and README/AGENTS instructions, respectively.

**Comment by Lewiscowles1986** [comment]
So I came in search of answers and this has been great after a small fix to a high visibility (to me) repo https://github.com/simonw/datasette-lite/pull/81

I think the approach taken to embrace the network is part of the failures seen in that project. Running around all over the internet, seems to me to be a very non-uniform experience for this type of app, so allowing for offline runs and looking for caching mechanisms, to me, seems a very natural fit. Thank you all
